### PR TITLE
Fixed release pets/mounts achievements to award only when fully earned and added un…

### PR DESCRIPTION
**Pull request by jerell-mendoza**
[//]: # (Note: See http://habitica.wikia.com/wiki/Using_Habitica_Git#Pull_Request for more info)

[//]: # https://github.com/HabitRPG/habitica/issues/8497#issuecomment-282994335

### Changes
[//]: # The current release pets, release mounts, and release both functions award achievements regardless of whether the user has any pets or mounts. There is now checking for each pet and mount before giving any of the achievements.

[//]: # 4b9dc32d-ad7d-4d2e-aaad-0543472f87f6

----
UUID:

### CODE REVIEW

File: test/common/ops/releaseBoth.js [hunk 1]
```diff
@@ -51,6 +51,15 @@ describe('shared.ops.releaseBoth', () => {
     expect(user.achievements.triadBingoCount).to.equal(2);
   });
 
+  it('does not grant triad bingo if any pet has not been previously found', () => {
+    let triadBingoCountBeforeRelease = user.achievements.triadBingoCount;
+    user.items.pets[animal] = -1;
+    let [, message] = releaseBoth(user);
```
Comment (paglias) [inline test/common/ops/releaseBoth.js]: can this be changed to `let message = releaseBoth(user)[1];`? I find it more readable

File: test/common/ops/releaseBoth.js [hunk 2]
```diff
@@ -51,6 +51,15 @@ describe('shared.ops.releaseBoth', () => {
...
+
+    expect(message).to.equal(i18n.t('mountsAndPetsReleased'));
+    expect(user.achievements.triadBingoCount).to.equal(triadBingoCountBeforeRelease);
+  });
+
   it('releases pets', () => {
     let [, message] = releaseBoth(user);
```
Comment (paglias) [inline test/common/ops/releaseBoth.js]: same here

**Review by paglias — CHANGES_REQUESTED** [review]

**Comment by paglias** [comment]
@jerellmendoodoo thanks for this! I've left two very small comments but overall it's okay

**Comment by jerell-mendoza** [comment]
Hi @paglias! Thanks for reviewing my code, I can rewrite those later tonight and update it.

**Comment by jerell-mendoza** [comment]
Hi @paglias, found a couple other tests in files releaseBoth, releasePets and releaseMounts - they all had the same way of assigning that variable for 'message' - I've modified them to match the way you recommended.

**Comment by paglias** [comment]
@jerellmendoodoo looks like some extra commits got included in the PR. Usually there's no need to update the code from `origin/develop` into your own branch unless there are conflicts but if you want to do it the best way is to rebase:

```
... pull develop in your branch from origin / develop ...
git checkout yourWorkingBranch
git rebase develop
```

and then force push (git push origin -f) to your branch.

Also usually it's better to work on a PR not from the `develop` branch but from a specific one that's only used for the PR. You can create one with `git checkout -b yourBranchName`

**Comment by jerell-mendoza** [comment]
@paglias ok thanks for the feedback, I'm pretty new to github and using git so it's all very foreign haha. But I didn't know you could just work on the PR branch. Do you have any recommendations as to where I can learn about rebasing and branching - I'm actually a little confused how that all works. Also, in regards to this issue - what should I do next since I've already made those commits?

**Comment by paglias** [comment]
@jerellmendoodoo for learing git I think this site is quite good https://git-scm.com/book/en/v2.

In this case, are you using the terminal for git commands?

**Comment by jerell-mendoza** [comment]
@paglias awesome thank you for that reference. And yes I'm using terminal
as of now.


On Wed, Mar 8, 2017 at 12:49 AM Matteo Pagliazzi <notifications@github.com>
wrote:

@jerellmendoodoo <https://github.com/jerellmendoodoo> for learing git I
think this site is quite good https://git-scm.com/book/en/v2.

In this case, are you using the terminal for git commands?

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/HabitRPG/habitica/pull/8539#issuecomment-284982615>,
or mute
the thread

**Comment by paglias** [comment]
Ok try doing this

```
git checkout develop
git checkout -b fix-release-pets-mounts
git fetch
git reset --hard upstream/develop
git cherry-pick 01b9412928fa7574ed6177d53c6664a6183a3804
git cherry-pick 6605711a415513da0c43dc151a83604b7af1a249
git cherry-pick fdbf9740d6f85742b05b712ca8ab0a15f0e4bec2
git cherry-pick d296ad7786154178c275f09465d2607d01b933da
git push origin fix-release-pets-mounts
```

This should create a new clean branch with only your commits, switch to it and push it github, then you should be able to create a new PR. I hope it's correct but I'm not 100% sure.

If it works correctly, but only if it works (otherwise you risk losing your work), then you can do

```
git checkout develop
git fetch
git reset --hard upstream/develop
git push --force origin develop
```

to clean up your develop branch.

To switch between branches you use `git checkout branchName` and to create a new one from the current checked out branch (and switch to it automatically) `git checkout -b newBranchName`

**Comment by jerell-mendoza** [comment]
Hi @paglias, I've followed your steps but for the cherry-pick I just picked the commits that were needed for the fix:
01b9412928fa7574ed6177d53c6664a6183a3804
6605711a415513da0c43dc151a83604b7af1a249
fdbf9740d6f85742b05b712ca8ab0a15f0e4bec2
(I think this was right). Then i pushed my branch. To double check, I downloaded the code in my branch and looked through the code and it looked correct and then ran through the different unit tests I created and that looked correct too.

Here's my new pull request:
https://github.com/HabitRPG/habitica/pull/8545

**Pull request closed without merging**